### ISSUE TITLE
Blurry text on Linux (lack of subpixel rendering?)

### ISSUE THREAD

**Issue by Heath123** [warpdotdev/Warp#4304]
### Discord username (optional)

circuit10

### Describe the bug

(make sure to view this image at 1:1 scale)
![image](https://github.com/warpdotdev/Warp/assets/13787163/9df0c6c7-d380-4802-8065-546ac427d506)
The text on the left from Warp is noticeably blurrier than the text on the right from GNOME Terminal. It looks like this might be due to not having subpixel rendering?
![image](https://github.com/warpdotdev/Warp/assets/13787163/57645adc-cc43-4252-9a42-667c87ff8a47)
The horizontal line on the "e" is also rendered worse, being smudged across two rows

### To reproduce

1. Run GNOME Terminal with these settings
![image](https://github.com/warpdotdev/Warp/assets/13787163/202f13e0-f1d1-4151-8324-1d518b5be7da)
2. Run Warp with these settings
![image](https://github.com/warpdotdev/Warp/assets/13787163/f94271c4-da0d-4191-9c5d-ef453e912012)
(any settings should work but I've tried to match them up somewhat for a more direct comparison)

### Expected behavior

The font should look crisp and be rendered properly

### Operating system

Linux

### Operating system and version

Ubuntu 23.10

### Shell Version

fish, version 3.6.1

### Current Warp version

v0.2024.02.20.08.01.stable_02

### Regression

No, this bug or issue has existed throughout my experience using Warp

### Does this block you from using Warp daily?

No

### Is this a Warp specific issue? (i.e. does it happen in Terminal, iTerm, Kitty, etc.)

Yes, this I confirmed this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b9d78064-c89e-4973-b153-5178a31ee54e

None

**Comment by mrvladus** [comment]
Same on Fedora GNOME. I have to use bigger font, but it's still blurry, not so much as smaller font though... Inteface icons look blurry too.

**Comment by dannyneira** [comment]
Hey folks! Thanks for your patience while trying Warp. Could you download this preview version of Warp and see if it fixes your issues:

[deb](https://releases.warp.dev/stable/v0.2024.02.27.08.01.stable_00/warp-terminal_0.2024.02.27.08.01.stable.00_amd64.deb)
[rpm](https://releases.warp.dev/stable/v0.2024.02.27.08.01.stable_00/warp-terminal-v0.2024.02.27.08.01.stable_00-1.x86_64.rpm)
[pacman](https://releases.warp.dev/stable/warp-terminal-v0.2024.02.27.08.01.stable_00-1-x86_64.pkg.tar.zst)
[appimage](https://releases.warp.dev/stable/v0.2024.02.27.08.01.stable_00/Warp-x86_64.AppImage)

If using Wayland, you can enable Wayland support by setting the `WARP_ENABLE_WAYLAND` envar (e.g`WARP_ENABLE_WAYLAND=1 warp-terminal`).

**Comment by mrvladus** [comment]
@dannyneira just tested on Wayland GNOME, still blurry.

**Comment by Heath123** [comment]
> Hey folks! Thanks for your patience while trying Warp. Could you download this preview version of Warp and see if it fixes your issues:
> 
> [deb](https://releases.warp.dev/stable/v0.2024.02.27.08.01.stable_00/warp-terminal_0.2024.02.27.08.01.stable.00_amd64.deb) [rpm](https://releases.warp.dev/stable/v0.2024.02.27.08.01.stable_00/warp-terminal-v0.2024.02.27.08.01.stable_00-1.x86_64.rpm) [pacman](https://releases.warp.dev/stable/warp-terminal-v0.2024.02.27.08.01.stable_00-1-x86_64.pkg.tar.zst) [appimage](https://releases.warp.dev/stable/v0.2024.02.27.08.01.stable_00/Warp-x86_64.AppImage)
> 
> If using Wayland, you can enable Wayland support by setting the `WARP_ENABLE_WAYLAND` envar (e.g`WARP_ENABLE_WAYLAND=1 warp-terminal`).

Thanks but this doesn't seem to fix it for me:
![image](https://github.com/warpdotdev/Warp/assets/13787163/2234f1b0-05fc-40fd-9634-16a8725102e7)

**Comment by dannyneira** [comment]
@yooakim there is a new release for Warp we pushed out yesterday. https://app.warp.dev/get_warp it has a number of bug fixes https://docs.warp.dev/getting-started/changelog#id-2024.02.29-v0.2024.02.27.08.01.stable_03

**Comment by yooakim** [comment]
Sad to report the issue is the same (for me) with the new release. Text looks blurry and is hard to read, at least the smaller sizes.

**Comment by desingh-rajan** [comment]
> Sad to report the issue is the same (for me) with the new release. Text looks blurry and is hard to read, at least the smaller sizes.

Same for me as well.

**Comment by thaddeuskrugman** [comment]
I'm having this problem on my Mac. Text on Wezterm looks nice and clear but warp looks a little fuzzy.

**Comment by tikonius** [comment]
Same issue here. Makes it very hard on the eyes to use NVIM. Otherwise Warp has been a great experience so far!

**Comment by joihn** [comment]
@dannyneira 
Issue seems to still be there

left, warp, blurry
right, terminator, not blurry
![image](https://github.com/warpdotdev/Warp/assets/11663917/6cf6bb23-d309-4de8-aa92-b689f29b7992)
**my ENV:**
ubuntu22, xwayland 
FullHD monitor
warp v0.2024.06.25.08.02.stable_01

**Comment by dannyneira** [comment]
Hey folks, thanks for your patience. I suspect it may be related to the native Wayland support we are working on or the font rendering library we currently use. No eta on a fix at this time, but as a workaround I recommend you try tweaking the enforce minimum contrast settings and the line height set to default or lower than 1.3. Also on Mac exclusively, there is the thin stokes setting which may help as well. https://docs.warp.dev/appearance/text-fonts-cursor

**Comment by alokedesai** [comment]
Hi everyone,
Appreciate your patience here. The latest version of Warp (v0.2026.04.01.08.39.stable_01) has an initial set of improvements to our text rendering backend on Linux / Windows. We're going to continue to invest in more improvements, but it should already be significantly better on low DPI monitors. 

Please try it out and let me know if you see any issues!

**Comment by avondalepark** [comment]
Wow, way too late for me but good luck. Can't believe you weren't taking Linux more seriously.

**Comment by alokedesai** [comment]
I'm going to close this out; if folks are still seeing issues please reopen or file a new issue. Thanks!

**Comment by rudrabhoj** [comment]
Text still look considerably more blurry to me, on both wayland and x11 so can't be a wayland specific bug. @alokedesai

**Comment by alokedesai** [comment]
@rudrabhoj Mind filing a new issue? Would also _love_ for folks to help contribute to make this better, we'd gladly accept contributions here.

**Comment by rudrabhoj** [comment]
Hey @alokedesai I am investigating it right now, not too far from a fix. Would open a PR for it soon :)

**Comment by rudrabhoj** [comment]
Made a lot of progress on this issue btw. Closely researched Zed's GPUI way of handling this problem, closest to our overall approach on warp. Current result (Left is warp vs Right is Ghostty, on FreeBSD-current+Niri+Helix Editor stack):

<img width="2560" height="1440" alt="Image" src="https://github.com/user-attachments/assets/0d0b5740-0452-4a11-999d-e8786d28bebf" />

Will test it more before opening a pr potentially. Might reduce glyph thickness a lil bit too.

**Comment by rudrabhoj** [comment]
<img width="2560" height="1440" alt="Image" src="https://github.com/user-attachments/assets/45ebb883-3e1d-4803-8418-19ee4503b741" />

<img width="2560" height="1440" alt="Image" src="https://github.com/user-attachments/assets/c2715188-984f-478f-a7ef-26933de25cce" />

More updates, after dozen or so more commits in my private branch I think I am able to get very close to how zed's font rendering pipeline worked for linux/bsd. 

First image is Warp vs Ghostty, second is Warp vs Zed, on same file.

**Comment by alokedesai** [comment]
@rudrabhoj Appreciate all of your hard work on this. Could you open a Pr with a tech spec? Would love to get a Warper to help out and align on the approach so we can get this merged!

**Comment by rudrabhoj** [comment]
Hey @alokedesai, I have opened two PRs actually. Look at PRs #9662 and warpdotdev/cosmic-text#9. #9662 can be tested right away, but to merge it we'd need cosmic text pr to be merged first as that is a dependency. In this PR I have matched the rendering quality of Zed, which is as good as it gets if we want to stay on Cosmic Text+Swash, it is almost as good as Freetype direction for most modern screens anyway. Freetype would be better for very low DPI+unhinted fonts. IMO rn effort of abandoning cosmic-text may not make sense at all so I stayed within the confines of Cosmic Text.

PS:
Temporarily #9662 refers to my fork of Warp's cosmic text 12, it is only for testing purposes. If we merge, we'd have to merge cosmic text pr first, as I mentioned before.

**Comment by rudrabhoj** [comment]
Hey @alokedesai did you get a chance to look at the prs?

**Comment by alokedesai** [comment]
Hey @rudrabhoj, sorry about that. We have an oz-agent that is set up to do the first rounds of review but i'm realizing that it's not setup correctly for the cosmic-text repo. 

I'll take a look at your PRs now and will also make sure we get our Oz OSS management app set up for any future PRs

**Comment by rudrabhoj** [comment]
Yes, problem with oz agent was it wasn't active on the cosmic-text repo, and cosmic-text repo was a blocker for the bot on main repo.
